### PR TITLE
Fix version code regex

### PR DESCRIPTION
**Pull request by godrei**
### Checklist

- [x] I've read and accepted the [Contribution Guidelines](https://github.com/bitrise-steplib/.github/blob/main/CONTRIBUTING.md)
- Requires _PATCH_ [version update](https://semver.org/)

### Context

This PR resolves version update-related bugs.

Resolves: https://github.com/bitrise-steplib/steps-change-android-versioncode-and-versionname/issues/11 and https://github.com/bitrise-steplib/steps-change-android-versioncode-and-versionname/issues/21

### Changes

- Separate versionCode and versionName update related functionality into `BuildGradleVersionUpdater` struct and `UpdateVersion` function.
- Add unit tests to reproduce version update bugs
- Updated versionCodeRegexPattern and versionNameRegexPattern

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -71,79 +71,112 @@ func failf(format string, v ...interface{}) {
...
+}
 
-	f, err := os.Open(cfg.BuildGradlePth)
-	if err != nil {
-		failf("Failed to read build.gradle file, error: %s", err)
-	}
+// UpdateResult stors the result of the version update.
+type UpdateResult struct {
+	NewContent                               string
+	FinalVersionCode, FinalVersionName       string
```
Comment (mateherber) [inline main.go:87]: I prefer all variables on new line for readibility

File: main.go [hunk 2]
```diff
@@ -71,79 +71,112 @@ func failf(format string, v ...interface{}) {
...
+	UpdatedVersionCodes, UpdatedVersionNames int
+}
 
-	var finalVersionCode, finalVersionName string
-	var updatedVersionCodes, updatedVersionNames int
+// UpdateVersion executes the version updates.
+func (u BuildGradleVersionUpdater) UpdateVersion(newVersionCode *int, versionCodeOffset int, newVersionName *string) (*UpdateResult, error) {
```
Comment (mateherber) [inline main.go:92]: Why pointers for `newVersionCode` and `newVersionName` my spider sense triggers.

Also why the pointer return value?

**Review by mateherber — CHANGES_REQUESTED** [review]

File: main.go (hunk 2 above)
Comment (godrei) [reply main.go]: Pointer arguments: initially I just wanted to move the update functionality into a new function, without modifying it. Updated the code to use value arguments.

Pointer return value: my intention was to use the `nil` return value if the update fails. I dropped this idea, as the function not just updating, but reading and exposing the versionCode and versionName (even if nothing was updated).

File: main.go [hunk 3]
```diff
@@ -71,6 +71,74 @@ func failf(format string, v ...interface{}) {
...
+		},
+
+		regexp.MustCompile(versionNameRegexPattern): func(line string, lineNum int, match []string) string {
+			oldVersionName := match[1]
+			res.FinalVersionName = oldVersionName
+			updatedLine := ""
+
+			if newVersionName != "" {
+				quotedNewVersionName := newVersionName
+				if !(strings.HasPrefix(quotedNewVersionName, `"`) && strings.HasSuffix(quotedNewVersionName, `"`)) {
```
Comment (mateherber) [inline main.go:121]: what is the suffix is missing? `"1.2.3`

**Review by mateherber — CHANGES_REQUESTED** [review]

File: main.go (hunk 3 above)
Comment (godrei) [reply main.go:121]: In this case, the step will insert an invalid versionName (`""1.2.3"`) and most likely Gradle will fail on such build.gradle file.

Comment (godrei) [reply main.go:121]: The intention was to remove trailing and leading double quotes, if any of them is missing, and add new leading and trailing `"`.
However in this commit: https://github.com/bitrise-steplib/steps-change-android-versioncode-and-versionname/pull/31/commits/acd4d23984c52ca304e1b4c7e50407a184bd4c7a I just removed the trailing unintentionally.

Fixed in the latest commit.

**Comment by godrei** [comment]
step.yml updated from this:

Screenshot 2021-04-09 at 14 28 19

to this:

Screenshot 2021-04-09 at 14 28 48

**Review by mateherber — APPROVED** [review]

**Pull request merged**